crates/rethnet_eth/src/remote.rs

```diff
@@ -37,6 +39,15 @@ where
     seq.end()
 }
 
+fn sequence_to_single<'de, D, T>(deserializer: D) -> Result<T, D::Error>
```

Comment:
I'm a bit confused about `single_to_sequence` and `sequence_to_single`. Why are they used in some methods and not in others? I can't see a pattern.

Reply:
Thanks for pointing this out. The pattern is _supposed to be_ that we use these helpers for every method that accepts only a single parameter, so that that parameter will be encoded/decoded as an array in the JSON. But I wasn't applying the pattern consistently, and there was an underlying bug/problem here because of that. I updated the tests to verify that there's always a sequence, in [this commit](https://github.com/NomicFoundation/hardhat/pull/3967/commits/7832956e8c2c669fb98069fe792c1a74089a9ab9), and you can see in the CI run for that commit [here](https://github.com/NomicFoundation/hardhat/actions/runs/5179422686/jobs/9332394337#step:5:185) that this test fails with the code as-is, and you can see how i fixed it (made the tests pass) in [this commit](https://github.com/NomicFoundation/hardhat/pull/3967/commits/7a4c1432507f26b0c01dc900040f95de4478147f).

Reply:
I still don't quite understand why the single parameter is a special case. All JSON-RPC parameters are inside an array, so having 0, 1, or more shouldn't be too different IMO.

For example, given these two methods:

```rust
    GetTransactionByHash(B256),
    GetTransactionCount(Address, BlockSpec),
```

In the first case the params are `["0x123..."]` and in the second case the params are something like `["0x123...", "latest"]`. In both cases you have an array in the JSON-RPC side, and in both cases you have an enum variant in the rust side. So I don't quite see why the single parameter is a special case. (Not saying that this is wrong, I'm probably missing something here, perhaps related to some serde's automagic behavior, but I'd like to understand it better.)

Reply:
It's because, by default, `serde` will serialize to an array only if it finds a `Vec`; and, if you tell it to deserialize to a data structure that isn't a `Vec`, and then it finds an array in the JSON, it will throw an error saying it didn't get what it expected. `single_to_sequence` makes `serde` serialize to a JSON array even though the source data type is _not_ a `Vec`. And `sequence_to_single` makes `serde` deserialize from an array even though our `enum` variant only told it to expect a single element.

Reply:
Also, for our `enum` variants that have multiple parameters (eg `GetTransactionCount(Address, BlockSpec)`), serde is automatically treating that list of parameters as a "sequence" to be encoded/decoded as an array in the JSON.

Reply:
> It's because, by default, `serde` will serialize to an array only if it finds a `Vec`; and, if you tell it to deserialize to a data structure that isn't a `Vec`, and then it finds an array in the JSON, it will throw an error saying it didn't get what it expected. `single_to_sequence` makes `serde` serialize to a JSON array even though the source data type is _not_ a `Vec`. And `sequence_to_single` makes `serde` deserialize from an array even though our `enum` variant only told it to expect a single element.

I guess I too strongly worded this. Everywhere I said `Vec` I guess I should have said "sequence", which could be a `Vec`, or alternatively could be a tuple, which covers the case of our enum variants with multiple parameters. (And maybe there are other forms of "sequence" as well? Not sure).

Reply:
Ah, ok, that makes sense. Especially this:

> Also, for our enum variants that have multiple parameters (eg GetTransactionCount(Address, BlockSpec)), serde is automatically treating that list of parameters as a "sequence" to be encoded/decoded as an array in the JSON.

So variants with a single value _are_ special (for serde).

Thanks for the explanation!